cheatsheets/Nodejs_security_cheat_sheet.md

```diff
@@ -0,0 +1,371 @@
+    if (toobusy()) {
+        // log if you see necessary
+        res.send(503, “Server Too Busy”);
+    } else {
+    next();
+    }
+});
+```
+
+## Perform application activity logging
+
+Logging application activity is an encouraged good practice. It makes it easier to debug any errors encountered during application runtime. It is also useful for security concerns, since it can be used during incident response. Also, these logs can be used to feed Intrusion Detection/Prevention Systems (IDS/IPS). In Node.js, there are some modules like Winston or Bunyan to perform application activity logging. These modules enable streaming and querying logs. Also, they provide a way to handle uncaught exceptions. With the following code, you can log application activities in both console and a desired log file.
+
+```JavaScript
+var logger = new (Winston.Logger) ({
+    transports: [
+        new (winston.transports.Console)(),
+        new (winston.transports.File)({ filename: ‘application.log’ })
+    ],
+    level: ‘verbose’
+});
+```
+
+Also, you can provide different transports so that you can save errors to a separate log file and general application logs to a different log file.
+
+## Perform input validation
+
+Input validation is a crucial part of application security. Input validation failures can result in many different types of application attacks. These include SQL Injection, Cross-Site Scripting, Command Injection, Local/Remote File Inclusion, Denial of Service, Directory Traversal, LDAP Injection and many other injection attacks. In order to avoid these attacks, input to your application should be sanitized first. The best input validation technique is to use a white list of accepted inputs. However, if this is not possible, input should be first checked against expected input scheme and dangerous inputs should be escaped. In order to ease input validation in Node.js applications, there are some modules like validator and mongo-express-sanitize. Also, you should escape all HTML and JavaScript content shown to users via application. You can use [escape-html](https://github.com/component/escape-html) or [node-esapi](https://github.com/ESAPI/node-esapi) libraries to perform output escaping.
+
+## Adhere to general application security principles
+
+This list has mainly focused on issues that are common in Node.js applications. Also, recommendations against these issues are given specific to Node.js environment. Apart from these, there are general principles that apply to web applications regardless of technologies used in application server. You should also keep those principles in mind while developing your applications. A very good reference document on these principles is developed and maintained by OWASP. You can always refer to [OWASP Web Application Security Testing Cheat Sheet](https://www.owasp.org/index.php/Web_Application_Security_Testing_Cheat_Sheet) to learn about vulnerabilities that may exist in your application.
```

Comment:
The cheatsheets on the wiki are deprecated, should point here instead: https://cheatsheetseries.owasp.org/